src/spec-node/imageMetadata.ts

```diff
@@ -102,14 +115,49 @@ const replaceProperties = [
 interface UpdatedConfigProperties {
 	customizations?: Record<string, any[]>;
 	entrypoints?: string[];
-	onCreateCommands?: (string | string[])[];
-	updateContentCommands?: (string | string[])[];
-	postCreateCommands?: (string | string[])[];
-	postStartCommands?: (string | string[])[];
-	postAttachCommands?: (string | string[])[];
+	onCreateCommands?: LifecycleCommand[];
+	updateContentCommands?: LifecycleCommand[];
+	postCreateCommands?: LifecycleCommand[];
+	postStartCommands?: LifecycleCommand[];
+	postAttachCommands?: LifecycleCommand[];
 	shutdownAction?: 'none' | 'stopContainer' | 'stopCompose';
 }
 
+export function lifecycleCommandOriginMapFromMetadata(metadata: ImageMetadataEntry[]): LifecycleHooksInstallMap {
+	const map: LifecycleHooksInstallMap = {
+		onCreateCommand: [],
+		updateContentCommand: [],
+		postCreateCommand: [],
+		postStartCommand: [],
+		postAttachCommand: [],
+		initializeCommand: []
+	};
+	for (const entry of metadata) {
+		const id = entry.id; // Only Features have IDs encoded in the metadata.
+		const origin = id ?? 'devcontainer.json';
+		for (const hook of pickFeatureLifecycleHookProperties) {
+			const command = entry[hook];
+			const substitutedCommand = substituteFeatureRoot(command, entry.featureRootFolder);
+			if (substitutedCommand) {
+				map[hook].push({ origin, command: substitutedCommand });
+			}
+		}
+	}
+	return map;
+}
+
+function mergeLifecycleHooks(metadata: ImageMetadataEntry[], hook: (keyof SchemaFeatureLifecycleHooks)): LifecycleCommand[] | undefined {
```

Comment:
Should this include the origin as above? That would change the JSON of the mergedConfiguration for `read-configuration`, so, e.g., Codespaces could use the origin in the UI when running the postAttachCommands.

Reply:
I've raised https://github.com/devcontainers/spec/pull/206 to change the devcontainer.json schema to be usable as the result of `mergedConfiguration`.  

After that, I think adding in a generic means to figure out the origin of a piece of the merged config would be useful (not just for lifecycle hooks, but for all the pieces.  That might be worth its own PR, so i'll leave it out here.

Reply:
And then if there's a generic way we cache the origin of config, we could remove the `lifecycleCommandOriginMapFromMetadata()` function above.